plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  context,
+  diag,
+  Span,
+  SpanAttributes,
+  SpanOptions,
+  SpanStatusCode,
+  trace,
+} from '@opentelemetry/api';
+import { getRPCMetadata, RPCType } from '@opentelemetry/core';
+import {
+  InstrumentationBase,
+  InstrumentationConfig,
+  InstrumentationNodeModuleDefinition,
+  safeExecuteInTheMiddle,
+} from '@opentelemetry/instrumentation';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import type { HookHandlerDoneFunction } from 'fastify/types/hooks';
+import type { FastifyInstance } from 'fastify/types/instance';
+import type { FastifyReply } from 'fastify/types/reply';
+import type { FastifyRequest } from 'fastify/types/request';
+import {
+  AttributeNames,
+  FastifyNames,
+  FastifyTypes,
+} from './enums/AttributeNames';
+import type { HandlerOriginal } from './types';
+import { safeExecuteInTheMiddleMaybePromise } from './utils';
+import { VERSION } from './version';
+
+export const ANONYMOUS_NAME = 'anonymous';
+
+/** Fastify instrumentation for OpenTelemetry */
+export class FastifyInstrumentation extends InstrumentationBase {
+  constructor(config: InstrumentationConfig = {}) {
+    super(
+      '@opentelemetry/instrumentation-fastify',
+      VERSION,
+      Object.assign({}, config)
+    );
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<any>(
+        'fastify',
+        ['^3.0.0'],
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Applying patch for fastify@${moduleVersion}`);
+          return this._patchConstructor(moduleExports);
+        },
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Removing patch for fastify@${moduleVersion}`);
+        }
+      ),
+    ];
+  }
+
+  private _hookOnRequest() {
+    const instrumentation = this;
+    return function onRequest(
+      request: FastifyRequest,
+      reply: FastifyReply,
+      done: HookHandlerDoneFunction
+    ) {
+      if (!instrumentation.isEnabled()) {
+        return done();
+      }
+      const rpcMetadata = getRPCMetadata(context.active());
+      const routeName = request.routerPath;
+      if (routeName && rpcMetadata?.type === RPCType.HTTP) {
+        rpcMetadata.span.updateName(`${request.method} ${routeName || '/'}`);
```

Comment:
this span already has all of those attributes, it is changing a name so it will be the same as we do with express

Reply:
The http span has the `http.route` attribute [set to `path` by default](https://github.com/open-telemetry/opentelemetry-js/blob/5fefbb274893a93033ef3786f75173c62682ade7/packages/opentelemetry-instrumentation-http/src/utils.ts#L456). But it should be set to the templated route [according to specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#http-server-semantic-conventions):
> http.route | string | The matched route (path template). | /users/:userID

For example:
```js
import fastify from 'fastify';
const app = fastify({logger: true});
app.addHook('onRequest', (req, reply, done) => {
    const authSuccess = false;
    if(!authSuccess) {
        reply.code(401).send();
    }
    done();
});
app.get('/:id', (req, reply) => {
    throw Error('some error from request handler');
});
app.listen(3000);
``` 

In the example above, the HTTP span will have the `http.route` attribute set to `/1234` where it should be set to `/:id` according to the spec. Since the hook terminated the request before handler is invoked, there will be no span in the generated trace with the route information at all. The only indication for this request's (templated) route will be set on the name of the HTTP span which is not suitable for backend processing pipelines IMO.

Reply:
fixed